modules/core/04-channel/keeper/upgrade.go

```diff
@@ -57,43 +61,222 @@ func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID strin
+	if !found {
+		return types.Upgrade{}, types.ErrSequenceSendNotFound
+	}
+	return types.Upgrade{
+		Fields:             fields,
+		Timeout:            upgradeTimeout,
+		LatestSequenceSend: seq - 1,
+	}, nil
+}
+
+// ChanUpgradeTry is called by a module to accept the first step of a channel upgrade handshake initiated by
+// a module on another chain. If this function is successful, the proposed upgrade will be returned. If the upgrade fails, the upgrade sequence will still be incremented but an error will be returned.
 func (k Keeper) ChanUpgradeTry(
 	ctx sdk.Context,
 	portID,
 	channelID string,
 	proposedConnectionHops []string,
-	upgradeTimeout types.Timeout,
+	proposedUpgradeTimeout types.Timeout,
 	counterpartyProposedUpgrade types.Upgrade,
 	counterpartyUpgradeSequence uint64,
 	proofCounterpartyChannel,
-	proofCounterpartyUpgrade []byte,
+	proofUpgrade []byte,
 	proofHeight clienttypes.Height,
 ) (types.Upgrade, error) {
-	// TODO
-	return types.Upgrade{}, nil
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	// the channel state must be in OPEN or INITUPGRADE if we are in a crossing hellos situation
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	// verify that the timeout set in UpgradeInit has not passed on this chain
+	if hasPassed, err := counterpartyProposedUpgrade.Timeout.HasPassed(ctx); hasPassed {
+		// abort here and let counterparty timeout the upgrade
+		return types.Upgrade{}, errorsmod.Wrap(err, "upgrade timeout has passed: ")
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return types.Upgrade{}, errorsmod.Wrapf(connectiontypes.ErrConnectionNotFound, "failed to retrieve connection on %s. Error: %s", channel.ConnectionHops[0], err.Error())
+	}
+
+	// make sure connection is OPEN
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return types.Upgrade{}, errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	// check the connection associated with the passed in connection hops by the chain sending the TryUpgrade
+	// matches the connection associated with the counterparty proposed upgrade
+	counterpartyHops := counterpartyProposedUpgrade.Fields.ConnectionHops
+	if !reflect.DeepEqual(counterpartyHops, proposedConnectionHops) {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidConnectionHops, "proposed connection hops (%s) does not match counterparty proposed connection hops (%s)", proposedConnectionHops, counterpartyHops)
+	}
```

Comment:
This doesn't seem to make sense to me. We should be looking up the new proposed connection using `proposedConnectionHops[0]`, and then asserting that its counterparty is equal to `counterpartyHops`

something like:

```go
connection, err := k.GetConnection(ctx, proposedConnectionHops[0])
if err != nil {
	return types.Upgrade{}, err
}

if counterpartyHops != connection.GetCounterparty().GetConnectionID() {
	return types.Upgrade{}, errorsmod.Wrap(...)
}
```

Reply:
true, good catch